File: BOJ/2000-3000/2805/2805.py
import sys
from collections import Counter

N,M=map(int,sys.stdin.readline().split())

trees=Counter(list(map(int,sys.stdin.readline().split())))

lo=0
hi=max(trees)

answer=0

while lo <= hi :
    mid=(lo+hi)>>1
    total=0

    for height,number in trees.items():
        if height > mid:
            total+=(height-mid)*number

    if total >= M :
        lo=mid+1
        if answer < mid:
            answer=mid
    else:
        hi=mid-1

print(answer)

File: BOJ/2000-3000/2805/2805.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main(){
    int N,M;
    int answer=0;

    vector<long long> trees;

    cin>>N>>M;
    
    for(int i=0;i<N;i++){
        long long tmp;
        cin>>tmp;
        trees.push_back(tmp);
    }

    long long lo=0;
    long long hi=*max_element(trees.begin(),trees.end());

    while(lo<=hi){
        long long mid=(lo+hi)/2;
        long long total=0;

        for(int i=0;i<N;i++)
            if(trees[i] > mid)
                total+=trees[i]-mid;
        
        if (total>=M){
            lo=mid+1;
            if (answer < mid)
                answer = mid;
        }
        else
            hi=mid-1;
    }
    cout<<answer<<endl;

    return 0;
}